Implementation with Redux Toolkit and Redux Saga

zain-sattar opened the pull request:
This pull request introduces the implementation of the Todo List app, using Redux Toolkit and Redux Saga.

Key Changes:

1) Integration of Redux Toolkit for state management
2) Implementation of Redux Saga for asynchronous operations.

zainsatar committed:
finish: a simple react todo list app using react query, react hook form and yup and json server

zain-sattar committed:
:art:Improve structure of the code

zain-sattar committed:
:art:Improve structure of the code

zain-sattar committed:
:art: improve structure of the code

zain-sattar committed:
:art: improve structure of the code

danialsyed6 committed:
Merge pull request #1 from zain-sattar/develop-ui

Develop and style the UI of todo list app

zain-sattar committed:
:art:
improve structure of the code

zain-sattar committed:
:sparkles: add new feature (edit a existing todo)

zain-sattar committed:
:art: improve structure of the code

zain-sattar committed:
:art: improve structure of the code

zain-sattar committed:
Resolved merge conflicts

zain-sattar committed:
:bug:
Fix a typo error in todoController.ts file

danialsyed6 committed:
Merge pull request #2 from zain-sattar/develop-functionality

Feature: Add Todo List Functionality (Add, Edit, Delete)

zain-sattar committed:
:sparkles: develop functionality of todo list app (add, remove, delete and edit) using redux toolkit and redux saga

danialsyed6 commented on src/components/TodoActions/index.tsx:
```diff
@@ -1,44 +1,94 @@
+import { RootState } from "../../store/store";
 
 function TodoActions() {
+  const selectedTodo = useSelector(
+    (state: RootState) => state.todoList.todoItem
+  );
+  const dispatch = useDispatch();
```
instead of useSelector and useDispatch use redux connect through container pattern.

danialsyed6 commented on src/components/TodoList/index.tsx:
```diff
@@ -1,36 +1,78 @@
+  const todos = useSelector((state: RootState) => state.todoList.todos);
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch({ type: "todos/fetchTodos" });
+  }, [dispatch]);
```
create an action for this.

danialsyed6 commented on src/store/store.ts:
instead of store/store.ts create store/index.ts file

danialsyed6 commented on src/store/todoSlice.ts:
```diff
@@ -0,0 +1,48 @@
+import { createSlice, PayloadAction } from "@reduxjs/toolkit";
+
+export interface Todo {
+  id?: number;
+  task: string;
+  isCompleted: boolean;
+}
+
+interface TodoState {
+  todos: Todo[];
+  todoItem: Todo | undefined;
+}
+
+const initialState: TodoState = {
+  todos: [],
+  todoItem: undefined,
+};
```
these should be in the todo/types.ts file

danialsyed6 commented on src/store/todoSlice.ts:
```diff
@@ -0,0 +1,48 @@
+    deleteTodo: (state, action: PayloadAction<number>) => {
+      state.todos = state.todos.filter((todo) => todo.id !== action.payload);
+    },
+    setTodoList: (state, action: PayloadAction<Todo[]>) => {
+      state.todos = action.payload;
+    },
+    setTodoItem: (state, action: PayloadAction<Todo | undefined>) => {
+      state.todoItem = action.payload;
+    },
```
these look like normal set-state functions. In case of setTodoList you should use getTodos which gets todos from the db. And in case of setTodoItem you should use getTodo whcih gets todo by id.

danialsyed6 commented on src/store/todoSlice.ts:
```diff
@@ -0,0 +1,48 @@
+  initialState,
+  reducers: {
+    addTodo: (state, action: PayloadAction<Todo>) => {
+      state.todos.push(action.payload);
```
you should never do mutable operations on the state. a reducer should always return a new value of state if it is changed.

danialsyed6 commented on src/store/todoSlice.ts:
```diff
@@ -0,0 +1,48 @@
+      state.todoItem = action.payload;
+    },
+    updateTodo: (state, action: PayloadAction<Todo>) => {
+      const { id, task, isCompleted } = action.payload;
+      const todo = state.todos.find((todo) => todo.id === id);
+      if (todo) {
+        todo.task = task;
+        todo.isCompleted = isCompleted;
+      }
+    },
```
this also looks like a mutable state update. try n find a better way.

danialsyed6 commented on src/store/todoSaga.ts:
```diff
@@ -0,0 +1,61 @@
+} from "./todoSlice";
+
+function* addTodoSaga(action: PayloadAction<Todo>): Generator<any, void, any> {
+  try {
+    yield call(addTodoApi, action.payload);
+  } catch (error) {
+    console.error("Error adding todo:", error);
+  }
```
instead of logging the error, you should have an error state in each slice and a fail action for each action. which in this case will be addTodoFail this should set the error state to true or an error message.

danialsyed6 reviewed (changes requested):
Try to strictly follow the folder structure in the following blog.
[https://itnext.io/scalable-redux-architecture-for-react-projects-with-redux-saga-and-typescript-f6afe1dece9b](url)

ahmadfaraz629 commented on src/store/todoSaga.ts:
```diff
@@ -0,0 +1,61 @@
+  }
+}
+
+function* editTodoSaga(action: PayloadAction<Todo>): Generator<any, void, any> {
+  try {
+    yield call(editTodoApi, action.payload);
+  } catch (error) {
+    console.error("Error fetching todos:", error);
+  }
+}
```
Give proper types instead of any

zain-sattar committed:
:art: updated code and folder structure

use connector pattern instead of useSelector and useDispatch
fix direct mutation issues in todo slice
updated the folder structure

zain-sattar committed:
:label: handle datatypes of variables and functions

ahmadfaraz629 commented on src/components/TodoActions/index.tsx:
```diff
@@ -1,51 +1,106 @@
+    else setValue("todoTask", null);
+  }, [selectedTodo, setValue]);
+
+  const formHeading = selectedTodo ? "Edit todo" : "New todo";
+  const submitButtonLabel = selectedTodo ? "Save" : "ADD TODO";
```
Create enums for this

zain-sattar committed:
:art: Improve code structure by introducing enums in todoList componnet

zain-sattar committed:
:art: Improve code structure by introducing enums in todoList componnet
